### PR TITLE
Expose certain layered repository structs to reuse in relish storage

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Part of https://github.com/zenithdb/zenith/pull/686 that touches layered repository and its surroundings.

### CODE REVIEW

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -240,7 +257,7 @@ impl LayeredRepository {
                 )?;
 
                 // List the layers on disk, and load them into the layer map
-                timeline.load_layer_map()?;
+                let _latest_timeline_on_disk = timeline.load_layer_map()?;
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:260]: This then gets sent to the upload queue in https://github.com/zenithdb/zenith/pull/686/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR254-R260

but to bring that code here, relish_storage interface requires changes which will bloat the PR hence omitted,

File: pageserver/src/layered_repository.rs [hunk 2]
```diff
@@ -1435,6 +1471,14 @@ impl LayeredTimeline {
         let metadata_path =
             LayeredRepository::save_metadata(self.conf, self.timelineid, self.tenantid, &metadata)?;
         if let Some(relish_uploader) = &self.relish_uploader {
+            let _local_timeline = LocalTimeline {
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:1474]: Again, here goes the download instead, no separate metadata file upload is happening: https://github.com/zenithdb/zenith/pull/686/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR1471-R1479

Leaf as is to not to drag in the relish_storage related changes.

File: pageserver/src/layered_repository.rs [hunk 3]
```diff
@@ -102,6 +103,22 @@ lazy_static! {
     .expect("failed to define a metric");
 }
 
+/// The name of the metadata file pageserver creates per timeline.
+pub const METADATA_FILE_NAME: &str = "metadata";
+
+/// All files and data, relevant to a certain pageserver timeline,
+/// found locally in its working directory.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LocalTimeline {
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:112]: The main reason to introduce this is to enforce storing images (aka certain layers in bulk), instead of manipulating disjoint layers.

Also, instead of operating file paths directly, we toss layer names and ids which is enough to construct us a file path in the moment we decide to upload/download the layer.

https://github.com/zenithdb/zenith/pull/686/files#diff-8ef63341d08bdfccc129283af160e1b6032defcd0bc929974510710cdf3d6478R57

File: pageserver/src/layered_repository.rs [hunk 4]
```diff
@@ -102,6 +103,22 @@ lazy_static! {
     .expect("failed to define a metric");
 }
 
+/// The name of the metadata file pageserver creates per timeline.
+pub const METADATA_FILE_NAME: &str = "metadata";
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:107]: Since I start to deal with metadata more and more, time to extract it into a separate constant at least.
Ideally, I think we should extract this + `filename.rs` contents + various layer name `path()` methods into some struct that manages all layered storage workdir access.
But that's a separate big task.

File: pageserver/src/layered_repository/delta_layer.rs [hunk 5]
```diff
@@ -367,7 +343,9 @@ impl Layer for DeltaLayer {
 }
 
 impl DeltaLayer {
-    fn path_for(
+    /// Gets the path in pageserver workdir, that should correspond
+    /// to a given metadata, interpreted as a delta layer.
+    pub fn path_for(
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository/delta_layer.rs:348]: This and the similar method in the image_layer.rs are used later during the upload-download phase to get the actual fs path of a certain relish: https://github.com/zenithdb/zenith/pull/686/files#diff-8ef63341d08bdfccc129283af160e1b6032defcd0bc929974510710cdf3d6478R608-R611

File: pageserver/src/layered_repository/filename.rs [hunk 6]
```diff
@@ -35,7 +37,7 @@ impl DeltaFileName {
     /// Parse a string as a delta file name. Returns None if the filename does not
     /// match the expected pattern.
     ///
-    pub fn from_str(fname: &str) -> Option<Self> {
+    pub fn parse_str(fname: &str) -> Option<Self> {
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository/filename.rs:40]: Clippy appealed that `from_str` might confuse people and make them think about `FromStr`.

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 7]
```diff
@@ -304,6 +300,12 @@ pub struct FreezeLayers {
     pub open: Option<Arc<InMemoryLayer>>,
 }
 
+/// A result of an inmemory layer data being written to disk.
+pub struct LayersOnDisk {
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository/inmemory_layer.rs:304]: The reason for this to appear is the need to have a clear separation between layers when decide what to upload.
When the whole relish storage sync loop starts, we need to scan both local and remote storages and compare the layers, to see which files are missing.
Part of that is https://github.com/zenithdb/zenith/pull/686/files#diff-8ef63341d08bdfccc129283af160e1b6032defcd0bc929974510710cdf3d6478R217

`Layer` trait does not have much info to use when doing that, opposed to `DeltaLayer` and `ImageLayer` structs ergo the change.

------

Also, this way we've got solid Rust code, no extra vtables and other sorcery happens now which is way better IMO.

File: pageserver/src/layered_repository.rs [hunk 8]
```diff
@@ -48,9 +48,9 @@ use zenith_utils::lsn::{AtomicLsn, Lsn, RecordLsn};
 use zenith_utils::seqwait::SeqWait;
 
 mod blob;
-mod delta_layer;
-mod filename;
-mod image_layer;
+pub mod delta_layer;
```
Comment (SomeoneToIgnore) [inline pageserver/src/layered_repository.rs:51]: Those will be used in the uploader to 
* hold the layer names in mem between uploads/downloads
* derive the actual FS path when the upload/download is carried out for a certain layer

https://github.com/zenithdb/zenith/pull/686/files#diff-8ef63341d08bdfccc129283af160e1b6032defcd0bc929974510710cdf3d6478R23-R25
(you might first need to expand `synced_storage.rs` file diff for the link to start to work)

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 7 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/inmemory_layer.rs:304]: This seems a bit fragile, or a leaky abstraction if you will. Currently, 'write_to_disk' returns zero/one image layer, and possibly some delta layers, but I consider that an implementation detail. It should be free to create any number of image and delta layers, if it thinks that it's useful.

You could call `filename()` and compare those. The file name is ultimately what we use to match up the files in S3 and on local disk. Or if dealing with filenames seems inconvenient, we could invent a new `LayerId` struct that contains all the same information that's currently in the filename.

> Also, this way we've got solid Rust code, no extra vtables and other sorcery happens now which is way better IMO.

Better how? Performance?

File: pageserver/src/layered_repository.rs (hunk 3 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:112]: I think we should introduce a new `LayerFileName` enum that can hold ImageFilename or DeltaFileName. The S3 code shouldn't need to distinguish between different kinds of layers.

The metadata path is always `tenants/{tenant_id}/timelines/{timeline_id}/metadata` , so you don't necessarily need a field for that, you can construct it from the tenant and timeline ids.

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 7 above)
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository/inmemory_layer.rs:304]: > Better how? Performance?

Yes, that + it feels harder to use various existential types, as they're not accepted everywhere, you cannot have multiple traits defined in a generic way, etc., etc.

We're lucky to not to use such things on the crate boundaries much, otherwise some subtle bugs as https://github.com/rust-lang/rust/issues/88438 may occur for us too.

Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository/inmemory_layer.rs:304]: > It should be free to create any number of image and delta layers, if it thinks that it's useful.

Sure thing, I've adjusted that to be a collection instead of Option now. I remember it was so before, but later some refactoring had changed it to be a single element and that confused me a bit.

> You could call filename() and compare those. The file name is ultimately what we use to match up the files in S3 and on local disk. Or if dealing with filenames seems inconvenient, we could invent a new LayerId struct that contains all the same information that's currently in the filename.

I like the idea and tried to use `PathBuf` everywhere and that almost worked. The biggest issue was the need to deal with LSNs for the remote timelines.
When we start, we have to read all timelines and
* determine the timeline with latest LSN to download
* react somehow on attempts to upload lower LSN layers

What are the ways to get the `disk_consistent_lsn` for a given remote timeline?
* read the metadata file
Doing that for every timeline is costly

* parse the file paths and get layer names with LSN data in them
That's what I do now, since I have to parse the paths anyway to extract `ZTenantId` and `ZTimelineId`anyway
Not sure if it's correct to derive the `disk_consistent_lsn` from this data, but simple to determine which timeline is the latest and download it, if needed.

* keep a special metadata file with latest `disk_consistent_lsn` and update it after every operation
Stas had proposed that, feels a bit scary IMO due to various concurrency reasons: we already fiddle with `metadata` file somewhat specially, by ensuring it's downloaded/uploaded as last + we still have to parse file paths to get the ids + that will bloat the code more

* something else??

So far I lean to operating the `LayerFileName` (https://github.com/zenithdb/zenith/pull/688/files#r720209536 feels like a very nice change to support that and in general) and lean to try Stas' proposal, but not in this PR series and rather when dealing with https://github.com/zenithdb/zenith/pull/689#issuecomment-931216193

File: pageserver/src/layered_repository.rs [hunk 9]
```diff
@@ -102,6 +103,21 @@ lazy_static! {
...
+/// All files and data, relevant to a certain pageserver timeline,
+/// found locally in its working directory.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LocalTimeline {
+    pub tenant_id: ZTenantId,
+    pub timeline_id: ZTimelineId,
+    pub has_metadata: bool,
+    pub layers: BTreeSet<LayerFileName>,
+    /// Latest Lsn that corresponds to the contents written to disk.
+    pub disk_consistent_lsn: Lsn,
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository.rs:118]: What is the value of `disk_consistent_lsn` and where do we get it if `has_metadata` is false?

Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:118]: `disk_consistent_lsn` helps upload/download loop to determine, whether a certain timeline should be uploaded or not.
Currently, checkpointer thread always tries to freeze the layers on disk, disregarding on whether there were actual updates to the timeline.

We get it in two places:
* [during freezing](https://github.com/zenithdb/zenith/pull/686/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR1467)
* [when loading timelines from disk during startup](https://github.com/zenithdb/zenith/pull/686/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR1080)

Now I start to think, that the latter should be changed, good catch (luckily, only the WIP branch should be updated to fix this).
`has_metadata` case does not matter much for uploads/downloads, as it gets filtered out, but looks like we can throw it away in this struct entirely, replacing with `pub disk_consistent_lsn: Option<Lsn>`

Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:118]: > we can throw it away in this struct entirely, replacing with `pub disk_consistent_lsn: Option<Lsn>`

Cool. That was exactly my thought.

File: pageserver/src/layered_repository.rs [hunk 10]
```diff
@@ -1432,10 +1469,16 @@ impl LayeredTimeline {
...
+
+        LayeredRepository::save_metadata(self.conf, self.timelineid, self.tenantid, &metadata)?;
+        if let Some(_relish_uploader) = &self.relish_uploader {
+            //     schedule_timeline_upload(LocalTimeline {
+            //         tenant_id: self.tenantid,
+            //         timeline_id: self.timelineid,
+            //         has_metadata: true,
+            //         layers: layer_uploads,
+            //         disk_consistent_lsn,
+            //     };
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository.rs:1481]: This change just disables the upload, right? Is it intentional?

File: pageserver/src/layered_repository.rs (hunk 9 above)
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:126]: Implemented and pushed.

File: pageserver/src/layered_repository.rs (hunk 10 above)
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:1566]: Yes, due to:
* this PR being a subset of a bigger https://github.com/zenithdb/zenith/pull/686 one. I'm cutting that WIP one to separate pieces, so we could discuss and merge them in hopefully comprehensible bits
* uploads are disabled by default and while worked somehow before, they were not super good at it with weird threads they've spawned. IMO, no big issue to "break" them temporarily for sake of merging a big WIP pr eventually, that improves things

Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:1566]: Thanks. It would be nice to have some comment explaining that.

Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:1566]: Sure, I can do that, but I think it's better to be done in the thread model-related PR: https://github.com/zenithdb/zenith/pull/689

since more relevant and it has to do the same disabling, but in a bigger scale.
So, ping me when you get there, this is what we have to discuss eventually :slightly_smiling_face:

File: pageserver/src/layered_repository.rs (hunk 9 above)
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs:126]: > Implemented and pushed.

Hmm. I don't see the change... Looks like you missed to git add something?

Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs]: Quite a juggling with branches I've cast on myself, sorry.
Pushed now + here's the "real" change in the  WIP branch with better context on how that change is treated during uploads: https://github.com/zenithdb/zenith/pull/686/commits/66a5bdaa359890ddd3cfa036a96dff6640d511d9

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 7 above)
Comment (lubennikovaav) [reply pageserver/src/layered_repository/inmemory_layer.rs:304]: >     * read the metadata file
>       Doing that for every timeline is costly

Is it? Metadata files should be very small and reading them once at pageserver start is not a big deal.

Moreover, any other way to get the `disk_consistent_lsn` seems to be incorrect,
or maybe I don't really understand the problem you're trying to solve here.

Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository/inmemory_layer.rs:304]: The problem can be vaguely formulated as "understand which layers we need to sync and how (download or upload)".
That has to be done from two directions:

* parse the local files and propagate the knowledge about them. Currently that's what pageserver does on init: now it unconditionally schedules the uploads of every startup-loaded timeline that has a proper metadata + pageserver uploads the frozen layers, covering all new files appearing
 
* parse the remote files, download some and keep the remote files with metadata in memory to compare with future uploads later

So, the main question is how to store the data in-memory to compare it later, and Heikki here proposed (if I've got him right) to directly store the `PathBuf` for every file, calculating the local path for remote files.
I almost like the idea, but the main culprit for me is the fact that I have to parse the remote file names anyway, to get the `ZTenantId` and `ZTimelineId`, so might make sense to parse the filenames also and ensure that the files are "correct".
Plus, I have to know some Lsn at least and I've used the one from file names to determine which timeline to download.

But the more I talk about the remote metadata files, the more I feel I have to parse it for every timeline and tenant after all unless we come up with something cleverer.

File: pageserver/src/layered_repository.rs [hunk 11]
```diff
@@ -1422,10 +1456,22 @@ impl LayeredTimeline {
...
+            for image_layer in new_historics.image_layers {
+                layer_uploads.push(image_layer.path());
+                historical_layers.push(Arc::new(image_layer));
+            }
+
             // If we created a successor InMemoryLayer, its predecessor is
             // currently the frozen layer. We need to update the predecessor
             // to be the latest on-disk layer.
-            if let Some(last_historic) = new_historics.last() {
+            if let Some(last_historic) = historical_layers.last() {
```
Comment (patins) [inline pageserver/src/layered_repository.rs:1474]: There's an important meaning to the last historical layer: it is the new predecessor.
This implementation gets it right, by doing the image_layer loop after the delta_layer one. It seems kind of brittle to me

File: pageserver/src/layered_repository.rs [hunk 12]
```diff
@@ -251,7 +263,18 @@ impl LayeredRepository {
                 )?;
 
                 // List the layers on disk, and load them into the layer map
-                timeline.load_layer_map(disk_consistent_lsn)?;
+                let loaded_map = timeline.load_layer_map(disk_consistent_lsn)?;
```
Comment (patins) [inline pageserver/src/layered_repository.rs:266]: I think the metadata is loaded twice (see just above this call + inside load_layer_map).
Also disk_consistent_lsn comes from the metadata, seems odd to use it and load it as well

File: pageserver/src/layered_repository.rs [hunk 13]
```diff
@@ -110,6 +110,18 @@ lazy_static! {
...
+/// The name of the metadata file pageserver creates per timeline.
+pub const METADATA_FILE_NAME: &str = "metadata";
+
+/// All files and data, relevant to a certain pageserver timeline,
+/// found locally in its working directory.
+#[derive(Debug)]
+struct LoadedLayerMap {
+    pub layers: Vec<PathBuf>,
+    /// Latest Lsn that corresponds to the contents written to disk.
+    pub disk_consistent_lsn: Option<Lsn>,
```
Comment (patins) [inline pageserver/src/layered_repository.rs:122]: When will this be false? If we can't find the metadata, should we just reject loading the layer map all together?

File: pageserver/src/layered_repository.rs [hunk 14]
```diff
@@ -1022,21 +1046,22 @@ impl LayeredTimeline {
...
         );
         let mut layers = self.layers.lock().unwrap();
-        let (imgfilenames, mut deltafilenames) =
-            filename::list_files(self.conf, self.timelineid, self.tenantid)?;
+        let timeline_files =
+            filename::list_timeline_files(self.conf, self.timelineid, self.tenantid)?;
 
         let timeline_path = self.conf.timeline_path(&self.timelineid, &self.tenantid);
-
+        let mut local_layers = Vec::with_capacity(timeline_files.image_layer_names.len());
```
Comment (patins) [inline pageserver/src/layered_repository.rs:1062]: Shouldn't this include delta layer len?

File: pageserver/src/layered_repository/filename.rs [hunk 15]
```diff
@@ -269,38 +306,52 @@ impl fmt::Display for ImageFileName {
...
 
+/// All files, found on the disk for the corresponding timeline.
+#[derive(Debug)]
+pub struct TimelineFiles {
+    pub image_layer_names: BTreeSet<ImageFileName>,
+    // Note that the order matters here, ergo [`BTreeSet`].
+    // The delta files are created in order starting from the oldest file,
+    // because each DeltaLayer needs a reference to its predecessor.
+    // [`DeltaFileName`] derives the ordering which is used here.
+    pub delta_layer_names: BTreeSet<DeltaFileName>,
```
Comment (patins) [inline pageserver/src/layered_repository/filename.rs:317]: Since there's no online querying, does it make more sense to just use Vecs and sort them before returning?

File: pageserver/src/layered_repository.rs (hunk 12 above)
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:266]: Oh my, how did I miss that. Thanks for spotting!

File: pageserver/src/layered_repository/filename.rs (hunk 15 above)
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository/filename.rs:317]: But circling back to https://github.com/zenithdb/zenith/pull/688/files#r723812773, looks like this part can stay the same, less new code is great, thanks.

File: pageserver/src/layered_repository.rs [hunk 16]
```diff
@@ -376,19 +387,21 @@ impl LayeredRepository {
 
         // fsync the parent directory to ensure the directory entry is durable
         if first_save {
-            let timeline_dir = File::open(&timeline_path)?;
-            timeline_dir.sync_all()?;
+            if let Some(timeline_path) = path.parent() {
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:390]: Why this change?

Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:390]: Mainly purism, can revert it back.

Before, `timeline_path` was constructed only to create the path to metadata file.
Later, `timeline_path` was used to sync the FS metadata after our metadata file was written.
So `timeline_path` was explicitly used as a parent, assuming that it never changes.

Now I have a method to get the metadata path directly and I can move away from the "metadata file is stored inside `timeline_path` directory" hardcoded knowledge.

I can do `.parent().unwrap()` as a middle ground to emphasize that we're dealing with correctly created metadata, but not fond of potential panics myself.

File: pageserver/src/layered_repository/filename.rs (hunk 15 above)
Comment (patins) [reply pageserver/src/layered_repository/filename.rs]: With the removal of predecessors in #717 order shouldn't matter anymore

File: pageserver/src/layered_repository.rs (hunk 4 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:114]: Maybe introduce this constant and the metadata_path function in a separate PR or commit, for easier review, even if you don't do any further refactoring right now?

File: pageserver/src/layered_repository.rs (hunk 13 above)
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs]: Looks like after I've dealt with double metadata loading issue, this struct is not needed, we just return the file paths and don't care about the Lsn.

File: pageserver/src/layered_repository.rs [hunk 17]
```diff
@@ -1028,9 +1037,10 @@ impl LayeredTimeline {
     }
 
     ///
-    /// Scan the timeline directory to populate the layer map
+    /// Scan the timeline directory to populate the layer map.
+    /// Gets all timeline-related files that were found and loaded.
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:1041]: I'd suggest `s/Gets/Returns/` for clarity

**Review by hlinnaka — APPROVED** [review]
LGTM now
